### ISSUE TITLE
Move SemaphoreCI tests to in-repo script

### ISSUE THREAD

**Issue by jodh-intel** [clearcontainers/runtime#168]
The SemaphoreCI setup currently has a lot of "build settings" (commands to run for the build).

All these commands should be moved into a script in the repository below `.ci/` so that the SemaphoreCI build settings can have a single command for setup, build and cleanup/teardown. The approach adopted for `cc-oci-runtime` seems like the best as those 3 scripts can be used for Travis, Semaphore or any other CI system we might wish to use:

- https://github.com/01org/cc-oci-runtime/blob/master/.travis.yml#L18..L25

**Comment by sboeuf** [comment]
I agree with this, we only want one call to each script for every step (Setup/Job1/After in Semaphore CI).

**Comment by jodh-intel** [comment]
Any update on this? `master` is currently broken due to an incorrect semaphore script (coupled to another bug admittedly). But, we should have all our CI scripts -- including where possible how to run them -- under version control.

**Comment by jodh-intel** [comment]
The problem btw seems to be that this semphore build step...

```
sudo -E PATH=$PATH make -C $GOPATH/src/github.com/clearcontainers/tests/ check
```

... should be setting `$GOPATH` to `$GOPATH/src/github.com/clearcontainers/tests/vendor/` before running `make check`. See:

- The build error:

  https://semaphoreci.com/clearcontainers/runtime/branches/master/builds/51

- The other related bug:

  https://github.com/clearcontainers/tests/pull/50

**Comment by jodh-intel** [comment]
Semaphore is now working. For reference, here are the scripts currently being run:

# Setup

```
sudo -E PATH=$PATH bash .ci/install_cc_runtime.sh
```

# Job 1

```
container_id=$(sudo docker create busybox /bin/sleep 60)
sudo docker ps -a
sudo docker start ${container_id}
sudo docker ps -a
sudo docker exec ${container_id} echo hello
sudo docker ps -a
sudo docker stop ${container_id}
sudo docker ps -a
sudo docker rm ${container_id}
sudo docker ps -a
sudo -E PATH=$PATH sh -c "cd $GOPATH/src/github.com/clearcontainers/tests && make check"
```

# After job

```
sudo cat /var/lib/clear-containers/runtime/runtime.log
sudo cat /var/log/upstart/cc-proxy.log
```

**Comment by chavafg** [comment]
I can take this issue if @sboeuf do not have a chance to do it. 
My proposal is:

1. use the CI scripts in the `clearcontainers/tests` repo and deprecate the ones here. 
2. On semaphore settings: 
    - Setup will only call `./setup.sh`
    - Job 1 will only call `./run.sh`
    - After job will only call `./teardown.sh`

If we want to modify something in the CI, we will change the scripts instead of the semaphore settings. 

`setup.sh` will clone `tests` repo and execute `.ci/setup.sh` from there. 
`run.sh` will contain what Job1 contains right now.
`teardown.sh` will contain what After Job contains right now.

**Comment by sboeuf** [comment]
@chavafg I am more than happy if you can handle this :)
I agree with what you will do, and that's the right approah IMO.
However there is one thing I would like to clarify. I think that we should have `setup-runtime.sh`, `run-runtime.sh` and `teardown-runtime.sh` for the runtime, `setup-virtcontainers.sh`, `run-virtcontainers.sh` and `teardown-virtcontainers.sh` for virtcontainers, and other scripts respectively for proxy and shim. Those steps could be different from a repo to another.

So basically in case of runtime for example, I think we can expect:
- Semaphore calls into `.ci/setup.sh` on the current repo (runtime in this case).
- This setup.sh knows it has to clone clearcontainers/tests repo, and it calls into `setup-runtime.sh`
- The same thing for `run` and `teardown`.

Does that make sense ?

**Comment by chavafg** [comment]
Assigned this item to myself. 

@sboeuf I think I am not following you. 

Why should the CI scripts be different for the other repos? 
We are going to execute all tests for all repos, right? If so, the scripts should be the same to setup the environment and to run the tests.

**Comment by sboeuf** [comment]
@chavafg Do you expect to have the same setup, same tests for every repo ?
I mean the shim setup is different (less steps) from the runtime setup. You don't need to setup as many things as you do with the whole runtime chain.
And you can imagine some very basic tests for the shim that have nothing in common with the runtime.

I imagine clearcontainers/tests as a database of all scripts that we need accross different repos. We want to reuse those scripts as much as possible. But I don't think we want the same setup, run and teardown for every repo, it does not make sense in my opinion.

@jodh-intel any thoughts ?

**Comment by jodh-intel** [comment]
@sboeuf - I'm slightly confused. I think you are saying that each system component that clones https://github.com/clearcontainers/tests will want to be able to call "component-specific" scripts that will live in https://github.com/clearcontainers/tests?

Couldn't we just pass a parameter to the scripts/Makefiles somehow so that the `tests` repo will do the right thing for whichever component it is testing (runtime, proxy, shim)?

So, the runtime's `ci-setup.sh` could be something like:

```
sudo -E PATH=$PATH sh -c "cd $GOPATH/src/github.com/clearcontainers/tests \
    && make check COMPONENT=runtime"
```

**Comment by sboeuf** [comment]
@jodh-intel Yes that's a good idea, no need to duplicate everything that way. But I really wanted to make sure that we were not expecting the same setup/run/teardown for every project.

**Comment by jodh-intel** [comment]
I think there might be some confusion here still; all I'm suggesting is that for any of our projects, we should only ever need to specify up to 3 scripts for any of our CI systems (setup, build, teardown). That makes it easy to, say, reconfigure Semaphore as you don't have to paste a ton on shell script into their web interface.

Those 3 scripts can have the same names in all the repos but I'm certainly not suggesting they will have the same contents ;)

**Comment by sboeuf** [comment]
Sounds good to me :)
